jaxopt/_src/anderson.py

```diff
@@ -0,0 +1,284 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Anderson Acceleration for fixed point in JAX."""
+
+from typing import Any
+from typing import Callable
+from typing import NamedTuple
+
+from dataclasses import dataclass
+
+import jax.numpy as jnp
+from jax.tree_util import tree_leaves, tree_structure
+
+from jaxopt._src import base
+from jaxopt._src import linear_solve
+from jaxopt._src.tree_util import tree_collapse, tree_uncollapse, tree_l2_norm
+
+
+class AndersonState(NamedTuple):
+  """Named tuple containing state information.
+  
+  Attributes:
+    iter_num: iteration number
+    value: pytree of current estimate of fixed point
+    error: residuals of current estimate
+    v_history: history of previous iterates
+    f_history: fixed_point_fun(v_history)
+  """
+  iter_num: int
+  value: Any
+  error: float
+  v_history: Any
+  f_history: Any
+
+
+@dataclass
+class AndersonAcceleration(base.IterativeSolver):
+  """Multi-dimensional fixed point solver using Anderson.
+
+  Attributes:
+    fixed_point_fun: a function ``fixed_point_fun(x, *args, **kwargs)``
+                     returning a pytree with the same structure and type as x
+                     each leaf must be an array (not a scalar)
+    history_length: size of history. Affect memory cost.
+    beta: momentum in Anderson updates. Default = 1.
+    maxiter: maximum number of iterations.
+    tol: tolerance (stoping criterion).
+    ridge: ridge regularization in solver.
+           Consider increasing this value if the solver returns 'NaN'.
+    verbose: whether to print error on every iteration or not.
+             Warning: verbose=True will automatically disable jit.
+    filter_leaves: boolean pytree with same structure as fixed_point_fun input,
+                   Will only take `True` leaves when computing the fixed point
+                   WARNING: this modifies optimality_fun to force zeros() on
+                            corresponding `False` leafs.
+                   Default = None (keep all leaves)
+    implicit_diff: whether to enable implicit diff or autodiff of unrolled
+      iterations.
+    implicit_diff_solve: the linear system solver to use.
+    jit: whether to JIT-compile the optimization loop (default: "auto").
+    unroll: whether to unroll the optimization loop (default: "auto")
+  """
+  fixed_point_fun: Callable = None
+  history_length: int = 5
+  beta: float = 1.
+  maxiter: int = 100
+  tol: float = 1e-5
+  ridge: float = 1e-5
+  filter_leaves: Any = None
+  verbose: bool = False
+  implicit_diff: bool = False
+  implicit_diff_solve: Callable = linear_solve.solve_normal_cg
+  jit: base.AutoOrBoolean = "auto"
+  unroll: base.AutoOrBoolean = "auto"
+
+  def tree_collapse(self, tree):
+    return tree_collapse(tree, self.filter_leaves)
+
+  def tree_uncollapse(self, tree, vec):
+    return tree_uncollapse(tree, vec, self.filter_leaves)
```

Comment:
I think these two methods are probably not necessary. The disadvantage of having them is that we have to scroll up to know what they do. If you decide to keep them, I would make them private methods (i.e. starting with a `_` prefix).